### PR TITLE
Add implementation fixes and parse output properly using JSON.mapping

### LINKED ISSUE #89: Expected Hash for #[]?(key : String), not Array(JSON::Type)
I found this :point_down: error today when I was working in my projects:

```
D, [2018-04-13 13:38:10 -05:00 #28884] DEBUG -- : workspace/didChangeConfiguration
E, [2018-04-13 13:39:46 -05:00 #28884] ERROR -- : A error was found while searching definitions
Expected Hash for #[]?(key : String), not Array(JSON::Type)
E, [2018-04-13 13:51:12 -05:00 #28884] ERROR -- : A error was found while searching definitions
Expected Hash for #[]?(key : String), not Array(JSON::Type)
E, [2018-04-13 13:51:40 -05:00 #28884] ERROR -- : A error was found while searching definitions
Expected Hash for #[]?(key : String), not Array(JSON::Type)
E, [2018-04-13 13:53:34 -05:00 #28884] ERROR -- : A error was found while searching definitions
Expected Hash for #[]?(key : String), not Array(JSON::Type)
E, [2018-04-13 13:56:44 -05:00 #28884] ERROR -- : A error was found while searching definitions
Expected Hash for #[]?(key : String), not Array(JSON::Type)
E, [2018-04-13 14:00:33 -05:00 #28884] ERROR -- : A error was found while searching definitions
Expected Hash for #[]?(key : String), not Array(JSON::Type)
E, [2018-04-13 14:00:33 -05:00 #28884] ERROR -- : A error was found while searching definitions
Expected Hash for #[]?(key : String), not Array(JSON::Type)
E, [2018-04-13 14:00:35 -05:00 #28884] ERROR -- : A error was found while searching definitions
Expected Hash for #[]?(key : String), not Array(JSON::Type)
E, [2018-04-13 14:03:10 -05:00 #28884] ERROR -- : A error was found while searching definitions
Expected Hash for #[]?(key : String), not Array(JSON::Type)
E, [2018-04-13 14:59:37 -05:00 #28884] ERROR -- : A error was found while searching definitions
Unexpected char 'E' at 1:1
```

I haven't seen this error before, I guess we need to review :point_down:  implementations feature again :sweat_smile: 

https://github.com/crystal-lang-tools/scry/blob/bf2660bd4a40574d72ff9cb4a0277ccabb3c56a6/src/scry/implementations.cr#L60-L61

### PR DESCRIPTION
**Pull request by faustinoaq**
Another fix :tada: 

This changes `JSON.parse` by `JSON.mapping`

Fixes #89 

**Edit:**

Add support for main file `.scry.cr` to customize implementation scope

### CODE REVIEW

File: src/scry/implementations.cr [hunk 1]
```diff
@@ -1,9 +1,26 @@
...
+require "./build_failure"
 
 module Scry
   # Using Crystal implementation to emulate GoTo Definition.
   struct Implementations
+    struct ImplementationsResponse
+      JSON.mapping(
+        status: String,
+        message: String,
+        implementations: {type: Array(ImplementationLocation), nilable: true}
```
Comment (bew) [inline src/scry/implementations.cr]: just a style comment, you can also do `implementations: Array(ImplementationLocation)?` (with a `?`), no opinion on that though, as you like

Comment (faustinoaq) [reply src/scry/implementations.cr]: @bew Done! Thank you :+1:

File: src/scry/implementations.cr [hunk 2]
```diff
@@ -49,15 +69,19 @@ module Scry
...
       response = crystal_tool(filename, position, scope)
       Log.logger.debug("response: #{response}")
-      parsed_response = JSON.parse(response)
-      implementations = parsed_response["implementations"]?
-      if implementations
-        implementations
-      else
+      response = (Array(BuildFailure) | ImplementationsResponse).from_json(response)
+      case response
+      when .is_a?(Array(BuildFailure))
```
Comment (bew) [inline src/scry/implementations.cr]: I think you can use `when Array(BuildFailure)` directly

Comment (faustinoaq) [reply src/scry/implementations.cr]: You're right, Done! :+1:

File: src/scry/implementations.cr [hunk 3]
```diff
@@ -49,15 +69,19 @@ module Scry
...
-      if implementations
-        implementations
-      else
+      response = (Array(BuildFailure) | ImplementationsResponse).from_json(response)
+      case response
+      when Array(BuildFailure)
         implementation_response
+      when ImplementationsResponse
+        if impls = response.implementations
+          impls
```
Comment (bew) [inline src/scry/implementations.cr]: why not always return a response by doing `response_with(impls)` here?
(and remove the case in `search`)

Comment (faustinoaq) [reply src/scry/implementations.cr]: Done!

**Pull request merged**